ssi-ldp/src/suites/mod.rs

```diff
@@ -2,18 +2,825 @@
-pub use tezos::*;
-pub use w3c::*;
+use tezos::*;
+#[cfg(feature = "w3c")]
+use w3c::*;
+
+use crate::{
+    prepare, prepare_nojws, sign, sign_nojws, use_eip712sig, use_epsig, verify, verify_nojws,
+    Error, LinkedDataDocument, LinkedDataProofOptions, Proof, ProofPreparation, ProofSuite,
+};
+
+use async_trait::async_trait;
+use std::{collections::HashMap as Map, str::FromStr};
+
+#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
+pub enum ProofSuiteType {
+    #[cfg(feature = "rsa")]
+    RsaSignature2018,
+    #[cfg(feature = "ed25519")]
+    Ed25519Signature2018,
+    #[cfg(feature = "ed25519")]
+    Ed25519Signature2020,
+    #[cfg(feature = "tezos")]
+    Ed25519BLAKE2BDigestSize20Base58CheckEncodedSignature2021,
+    #[cfg(feature = "tezos")]
+    P256BLAKE2BDigestSize20Base58CheckEncodedSignature2021,
+    #[cfg(feature = "secp256k1")]
+    EcdsaSecp256k1Signature2019,
+    #[cfg(feature = "secp256k1")]
+    EcdsaSecp256k1RecoverySignature2020,
+    #[cfg(feature = "eip")]
+    Eip712Signature2021,
+    #[cfg(feature = "eip")]
+    EthereumPersonalSignature2021,
+    #[cfg(feature = "eip")]
+    EthereumEip712Signature2021,
+    #[cfg(feature = "tezos")]
+    TezosSignature2021,
+    #[cfg(feature = "tezos")]
+    TezosJcsSignature2021,
+    #[cfg(feature = "solana")]
+    SolanaSignature2021,
+    #[cfg(feature = "aleo")]
+    AleoSignature2021,
+    #[cfg(feature = "w3c")]
+    JsonWebSignature2020,
+    #[cfg(feature = "secp256r1")]
+    EcdsaSecp256r1Signature2019,
+    CLSignature2019,
+    #[cfg(feature = "test")]
+    NonJwsProof,
+    #[cfg(feature = "test")]
+    #[serde(rename = "ex:AnonCredPresentationProofv1")]
+    AnonCredPresentationProofv1,
+    #[cfg(feature = "test")]
+    #[serde(rename = "ex:AnonCredDerivedCredentialv1")]
+    AnonCredDerivedCredentialv1,
+}
+
+// #[derive(Debug, Error)]
+// #[error(transparent)]
+// pub struct ParseProofSuiteError(#[from] ErrorRepr);
+impl FromStr for ProofSuiteType {
+    type Err = serde_json::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        serde_json::from_value(json!(format!("{s}")))
+    }
+}
+
+pub enum SignatureType {
+    JWS,
+    LD,
+}
+
+impl ProofSuiteType {
+    pub fn signature_type(&self) -> SignatureType {
+        match self {
+            #[cfg(feature = "rsa")]
+            Self::RsaSignature2018 => SignatureType::JWS,
+            #[cfg(feature = "ed25519")]
+            Self::Ed25519Signature2018 => SignatureType::JWS,
+            #[cfg(feature = "ed25519")]
+            Self::Ed25519Signature2020 => SignatureType::LD,
+            #[cfg(feature = "tezos")]
+            Self::Ed25519BLAKE2BDigestSize20Base58CheckEncodedSignature2021 => SignatureType::JWS,
+            #[cfg(feature = "tezos")]
+            Self::P256BLAKE2BDigestSize20Base58CheckEncodedSignature2021 => SignatureType::JWS,
+            #[cfg(feature = "secp256k1")]
+            Self::EcdsaSecp256k1Signature2019 => SignatureType::JWS,
```

Comment:
Is it possible to be able to have either a JWT VC or a LD Proof for the same proof suite?

Reply:
yeah for at least some of them, [here](https://w3c-ccg.github.io/ld-cryptosuite-registry/#ecdsasecp256k1signature2019) secp256k1-2019 is listed in the LD proof suites as is the JsonWebSig, there are probably more examples. Unfortunately I'm not sure where to look for a canonical list of JWT-supporting suites, or if there is one

Reply:
Perhaps some dual-use `ProofSuiteType` variants can hold a `SignatureType` internally and this fn can either return the internal value (if there is one) or return the supported constant one, e.g.
```
            #[cfg(feature = "tezos")]
            Self::P256BLAKE2BDigestSize20Base58CheckEncodedSignature2021 => SignatureType::JWS,
            #[cfg(feature = "secp256k1")]
            Self::EcdsaSecp256k1Signature2019(t) => t,
```

Reply:
I see, seems like a tricky thing to solve correctly. I guess it would have to be a field in `LinkedDataProofOptions`. And for the verification I've been thinking of making the `Proof` type an enum to make sure you can't have a `jws` and a `proof_value`, and then we can separate the proof types/formats, but this probably incurs its own set of problems.

Should we make an issue and address it later, considering this has always been the behaviour?

Reply:
LDPOpts would also be fine and a proof enum would be neat, but yeah I think it would be ok to address later, these are basically the common usage cases already.

Reply:
#491